pkg/controller/memberstatus/memberstatus_controller.go

```diff
@@ -0,0 +1,305 @@
+// It then takes the status from the cluster object and adds it to MemberStatus. Finally, it checks its status and will return an error if
+// its status is not ready
+func (r *ReconcileMemberStatus) hostConnectionHandleStatus(reqLogger logr.Logger, memberStatus *toolchainv1alpha1.MemberStatus) error {
+	// look up host connection status
+	fedCluster, ok := r.getHostCluster()
+	if !ok {
+		notFoundMsg := errMsgHostConnectionNotFound
+		notFoundReason := reasonHostConnectionNotFound
+		memberStatus.Status.HostConnection = kubefed_v1beta1.KubeFedClusterStatus{
+			Conditions: []kubefed_v1beta1.ClusterCondition{
+				{
+					Type:          "Ready",
+					Status:        corev1.ConditionFalse,
+					Reason:        &notFoundReason,
+					Message:       &notFoundMsg,
+					LastProbeTime: metav1.Now(),
+				},
+			},
+		}
+		return fmt.Errorf(notFoundMsg)
+	}
+	memberStatus.Status.HostConnection = *fedCluster.ClusterStatus.DeepCopy()
+
+	// check conditions of host connection
+	if !util.IsClusterReady(fedCluster.ClusterStatus) {
+		return fmt.Errorf("the host connection is not ready")
+	}
+
+	return nil
+}
+
+// memberOperatorHandleStatus retrieves the Deployment for the member operator and adds its status to MemberStatus. It returns an error
+// if any of the conditions have a status that is not 'true'
+func (r *ReconcileMemberStatus) memberOperatorHandleStatus(reqLogger logr.Logger, memberStatus *toolchainv1alpha1.MemberStatus) error {
+	operatorStatus := toolchainv1alpha1.MemberOperatorStatus{
+		Version:        version.Version,
+		Revision:       version.Commit,
+		BuildTimestamp: version.BuildTime,
+	}
+
+	// look up status of member deployment
+	memberOperatorDeploymentName, err := getMemberOperatorDeploymentName()
+	if err != nil {
+		err = errs.Wrap(err, errMsgCannotGetDeployment)
+		errCondition := newComponentErrorCondition(reasonNoDeployment, err.Error())
+		operatorStatus.Conditions = []toolchainv1alpha1.Condition{*errCondition}
+		memberStatus.Status.MemberOperator = operatorStatus
+		return err
+	}
+	memberDeploymentName := types.NamespacedName{Namespace: memberStatus.Namespace, Name: memberOperatorDeploymentName}
+	memberDeployment := &appsv1.Deployment{}
+	err = r.client.Get(context.TODO(), memberDeploymentName, memberDeployment)
+	if err != nil {
+		err = errs.Wrap(err, errMsgCannotGetDeployment)
+		errCondition := newComponentErrorCondition(reasonNoDeployment, err.Error())
+		operatorStatus.Conditions = []toolchainv1alpha1.Condition{*errCondition}
+		memberStatus.Status.MemberOperator = operatorStatus
+		return err
+	}
+
+	// get and check conditions of member deployment
+	conditionsReady := true
+	deploymentConditions := []appsv1.DeploymentCondition{}
+	for _, condition := range memberDeployment.Status.Conditions {
+		deploymentConditions = append(deploymentConditions, condition)
+		conditionsReady = conditionsReady && condition.Status == corev1.ConditionTrue
+	}
```

Comment:
It’s possible to get a state when the deployment can’t roll out a new version but the old/current one is still running.
We should make sure we cover that case.
Rajiv, can you double check that?

And, no, if there is no operator pods running at all (even some old version) then there is no controller to check the status and it will be caught by the host status.

Reply:
> It’s possible to get a state when the deployment can’t roll out a new version but the old/current one is still running.

If there is a problem with starting the pod, then yes, you are rigth. 
Actually, I've been playing with it a bit and I observed that when I release a new version of the operator and delete the image from the repository (which causes `ImagePullBackOff` error while doing the update) then I don't see any error in any of the conditions in the `Deployment` yaml file. The only thing that shows that there is a problem is the number of the unavailable replicas:
```yaml
status:
  observedGeneration: 5
  replicas: 2
  updatedReplicas: 1
  readyReplicas: 1
  availableReplicas: 1
  unavailableReplicas: 1
  conditions:
    - type: Available
      status: 'True'
      lastUpdateTime: '2020-06-29T14:11:58Z'
      lastTransitionTime: '2020-06-29T14:11:58Z'
      reason: MinimumReplicasAvailable
      message: Deployment has minimum availability.
    - type: Progressing
      status: 'True'
      lastUpdateTime: '2020-06-29T14:12:03Z'
      lastTransitionTime: '2020-06-29T14:11:49Z'
      reason: ReplicaSetUpdated
      message: ReplicaSet "host-operator-684877b4f8" is progressing.
```

> And, no, if there is no operator pods running at all (even some old version) then there is no controller to check the status and it will be caught by the host status.

Yes, this is what I was pointing to - a case when there is some panic in the member-operator after the old replica was replaced. So, I'm curious - if we want to check the `Deployment`/`Pods` of the member operator from the host anyway, why don't we check everything there? Then we would have only one place (in host operator) where we check all member operator deployments. And vice versa, we could check from the member operator the deployment of the host operator.

Reply:
In addition, if I break the deployment by deleting the ServiceAccount - which is the most frequest issue we've been facing for now - then there is no pod running of course. I believe that we should take care of this case for both operators - host and member.

Reply:
> if we want to check the Deployment/Pods of the member operator from the host anyway

We don't want to check member operator deployments/pods from the host. Member deployments/pods/etc are  implementation details of the member status thing. Host status will only make sure the member is READY **and** the update time of the status is not too old. If there is no controller to update the member status then the update time will be outdated.

But I wonder if we want to rename "updateTime" to "lastProbeTime" so it's obvious what this field for.

> In addition, if I break the deployment by deleting the ServiceAccount - which is the most frequest issue we've been facing for now - then there is no pod running of course. I believe that we should take care of this case for both operators - host and member.

It's covered by the updateTime/lastProbeTime timestamp too.

Reply:
+1 I like the idea of `lastProbeTime`. I'll make the change to the api as part of the status changes on the host side and will update the member as well.

Reply:
fair enough :+1: 
btw, shouldn't we check also the `unavailableReplicas` field in the Deployment status?

Reply:
I updated it to now check `unavailableReplicas` and report an not ready status in that case as well

```
status:
  conditions:
  - lastTransitionTime: "2020-06-30T10:03:58Z"
    lastUpdatedTime: "2020-06-30T10:03:58Z"
    message: 'components not ready: [memberOperator]'
    reason: ComponentsNotReady
    status: "False"
    type: Ready
  hostConnection:
    conditions:
    - lastProbeTime: "2020-06-30T10:03:48Z"
      lastTransitionTime: "2020-06-30T10:02:48Z"
      message: /healthz responded with ok
      reason: ClusterReady
      status: "True"
      type: Ready
    region: ""
  memberOperator:
    buildTimestamp: "2020-06-30T20:14:04Z"
    conditions:
    - lastTransitionTime: "2020-06-30T10:03:58Z"
      lastUpdatedTime: "2020-06-30T10:03:58Z"
      message: the member operator deployment has unavailable replicas
      reason: DeploymentUnavailableReplicas
      status: "False"
      type: Ready
    deployment:
      name: member-operator
    revision: e38ad2d0883e3295ae32a26d96c04dfc7a9cb57c-dirty
    version: 0.0.1
```